build-tools/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java

```diff
@@ -192,39 +209,56 @@ private static void configureClaimClustersHook(Gradle gradle, TestClustersRegist
-                    if (task instanceof TestClustersAware == false) {
-                        return;
-                    }
-                    // we only start the cluster before the actions, so we'll not start it if the task is up-to-date
-                    TestClustersAware awareTask = (TestClustersAware) task;
-                    awareTask.beforeStart();
-                    awareTask.getClusters().forEach(registry::maybeStartCluster);
-                }
-
-                @Override
-                public void afterActions(Task task) {}
+        private void configureStartClustersHook(
+            Gradle gradle,
+            TestClustersRegistry registry,
+            Provider<TaskEventsService> testClusterTasksService
+        ) {
+            testClusterTasksService.get().registry(registry);
+            gradle.getTaskGraph().whenReady(taskExecutionGraph -> {
+                taskExecutionGraph.getAllTasks()
+                    .stream()
+                    .filter(task -> task instanceof TestClustersAware)
+                    .map(task -> (TestClustersAware) task)
+                    .forEach(awareTask -> {
+                        testClusterTasksService.get().register(awareTask.getPath(), awareTask);
+                        awareTask.doFirst(task -> {
+                            awareTask.beforeStart();
+                            awareTask.getClusters().forEach(registry::maybeStartCluster);
+                        });
+                    });
             });
         }
+    }
+
+    static public abstract class TaskEventsService implements BuildService<BuildServiceParameters.None>, OperationCompletionListener {
+
+        Map<String, TestClustersAware> tasksMap = new HashMap<>();
+        private TestClustersRegistry registryProvider;
+
+        public void register(String path, TestClustersAware task) {
+            tasksMap.put(path, task);
+        }
 
-        private static void configureStopClustersHook(Gradle gradle, TestClustersRegistry registry) {
-            gradle.addListener(new TaskExecutionListener() {
-                @Override
-                public void afterExecute(Task task, TaskState state) {
-                    if (task instanceof TestClustersAware == false) {
-                        return;
-                    }
+        public void registry(TestClustersRegistry registry) {
+            registryProvider = registry;
+        }
+
+        @Override
+        public void onFinish(FinishEvent finishEvent) {
+            if (finishEvent instanceof TaskFinishEvent taskFinishEvent) {
+                // Handle task finish event...
+                String taskPath = taskFinishEvent.getDescriptor().getTaskPath();
+                if (tasksMap.containsKey(taskPath)) {
+                    TestClustersAware task = tasksMap.get(taskPath);
                     // always unclaim the cluster, even if _this_ task is up-to-date, as others might not have been
```

Comment:
I don't think the comment makes sense anymore. Since we're registering clusters in a `doFirst` by definition, if they exist in this map they _must_ have executed.